File: openfeed-frontend/actions/reranker.ts
interface RerankedDocument {
  relevance_score: number;
  index: number;
}

export async function rerankTexts(
  query: string,
  documents: string[],
): Promise<RerankedDocument[]> {
  const response = await fetch("https://api.voyageai.com/v1/rerank", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${process.env.VOYAGEAI_API_KEY || ""}`,
    },
    body: JSON.stringify({
      query,
      documents,
      model: "rerank-2.5-lite",
    }),
  });

  const json = await response.json();

  if (!response.ok || !json.data || !Array.isArray(json.data)) {
    console.error("Reranking API error:", json);
    return [];
  }

  return json.data.map(
    (d: any) =>
      ({
        relevance_score: d.relevance_score,
        index: d.index,
      }) as RerankedDocument,
  );
}

File: openfeed-backend/openfeed/reranker.py
import requests
from pydantic import BaseModel

from openfeed.config import settings


class RerankedDocument(BaseModel):
    relevance_score: float
    index: int


def rerank(query: str, documents: list[str]):
    response = requests.post(
        "https://api.voyageai.com/v1/rerank",
        headers={
            "Authorization": f"Bearer {settings.voyageai_api_key}",
        },
        json={"query": query, "documents": documents, "model": "rerank-2.5-lite"},
        timeout=(10, 30),
    )
    data = response.json()["data"]
    return [RerankedDocument.model_validate(d) for d in data]
